### DISCUSSION TITLE
How is py_binary defined?

### DISCUSSION THREAD

**Discussion by allsey87** [bazel-contrib/rules_python#2036] [Q&A] [answered]
I am a bit confused about how `py_binary` is defined. It seems that you can use this function in Bazel without importing anything, but it is also possible to override this default provided `py_binary` with a version from `rules_python` by `load`ing it in. Does `py_binary` refer to different implementations in these two cases? If so, where are the differences between these implementations documented?

**Comment by rickeylev** [comment, marked as answer]
You've pretty much got it.

Bazel provides an implementation of the core python rules. However, they are being moved out of Bazel and into rules_python. Rules_python is a drop in replacement for the old rules built into bazel. The two implementations have almost identical behavior. There's not many differences and they are pretty minor differences (rules_python probably has a few fixes and features beyond it, actually). There's an issue in the Bazel repo that has a list of differences, but I'm on mobile and can't find it.

Eventually, Bazel will remove the built-in implementations. I'm not sure when that'll happen. Hopefully Bazel 8.

**Reply by allsey87** [reply]
I think this is the issue that you were referencing: https://github.com/bazelbuild/bazel/issues/15897
